Singularize like pluralize

matrinox opened the pull request:
### Summary

- Makes `String#singularize` similar to `String#pluralize` by adding a `count` argument.
- Adds new `String#with_count` which is like a combination of singularize and pluralize. It makes working with words easier by eliminating the need to know what plurality it was originally.

matrinox committed:
Make singularize like pluralize with count

------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
On branch singularize-like-pluralize - Wed 23 Aug 2017 17:02:02 PDT by matrinox <geofflee25@gmail.com>
Amended commit - Wed 23 Aug 2017 17:13:53 PDT by matrinox <geofflee25@gmail.com>

matrinox committed:
Add String#with_count

It simplifies use of singularize and pluralize for words you don't know the plurality of
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
On branch singularize-like-pluralize - Wed 23 Aug 2017 17:28:17 PDT by matrinox <geofflee25@gmail.com>

rafaelfranca commented:
Implementation looks good but I'm not sure those methods would be useful.

@dhh thoughts?

rafaelfranca commented:
Implementation looks good but I'm not sure those methods would be useful.

@dhh thoughts?

matthewd commented:
The inflector tries its best, but generally doesn't do particularly well when asked to pluralize an already-plural word / singularize an already-singular. I don't think we should add new API that relies on a belief it'll work correctly.

As for adding a count to singularize.. I can see the appeal of symmetry, but `.singularize(count)` just sounds peculiar to me -- I think it's the weighting between the common and exceptional cases.

matrinox commented:
@matthewd You can make the argument that `.pluralize(count)` has the same issues as `.singularize(count)`: they both have cases where it doesn't do what its method name says it should.

> The inflector tries its best, but generally doesn't do particularly well when asked to pluralize an already-plural word / singularize an already-singular. I don't think we should add new API that relies on a belief it'll work correctly.

For me, the use case are with words that you don't know the plurality of. If you did, you would just use `singularize` or `pluralize` depending on if it's plural or singular, respectively. There may be an issue but it's still better than just using pluralize and having it now singularize when you pass `1` to it because all that does is duplicate the string.

igbanam commented on activesupport/lib/active_support/core_ext/string/inflections.rb:
```diff
@@ -52,9 +56,49 @@ def pluralize(count = nil, locale = :en)
+  #   'octopi'.with_count(2)           # => "octopus"
+  #   'sheep'.with_count(2)            # => "sheep"
+  #   'word'.with_count(2)             # => "word"
+  #   'the blue mailmen'.with_count(2) # => "the blue mailman"
```
This is a bit confusing. How does pluralizing "The blue mailmen" end up as "The blue mailman"? So… two blue *mailmen(s)* is a blue mailman? What am I missing?

matrinox replied:
I copied the examples above and hastily changed them. Thanks for the catch :)

dhh commented:
I'm probably just thick, but I don't understand the usecase here. When are you going to be calling #singularize(2)? Why is that better than calling #pluralize? If it's because you don't know the count up front, then I think calling it explicitly with `string.send(count.many? ? :pluralize : :singularize)` or whatever is clearer.

matrinox commented:
@dhh The issue is that `pluralize(1)` doesn't actually singularize the string. So if you have a string that is already plural (like "days"), you're forced to use `singularize`. Except it doesn't accept a count. This Pr just makes it symmetrical.

The use case is if you know the string is plural, use `.singularize`, if the string is singular, use `.pluralize`

matrinox committed:
Fix the comments for #with_count

------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
On branch singularize-like-pluralize - Thu 24 Aug 2017 13:07:24 PDT by matrinox <geofflee25@gmail.com>

dhh commented:
Ah, I just saw that String#pluralize takes a count. That's imo a mistake. I know it's an old mistake (the commit is from 2011!), but it's still a mistake. I'd rather deprecate that form and get rid of it.

I am sympathetic to the idea of taking a singular word and coming up with a new method name that can potentially pluralize it based on a count. But it wouldn't be through a count argument that makes pluralize singularize or vice versa. I don't think that's a good API at all.

Not liking the #with_count method name for this. Too ambiguous. Also, I'd seriously question the need for a method like that to be mixed into String directly, rather than being part of the Inflector.

dhh commented:
I did a quick search and it doesn't seem like there's a single word that represents the process of either singularizing or pluralizing a word. So we could just invent one, like `#plugularize`. But better if we can find an existing word that's closer. Then it can take the place of the #with_count method you have and we can drop the singularize(count) and deprecate pluralize(count).

matrinox commented:
> Ah, I just saw that String#pluralize takes a count. That's imo a mistake.

@dhh I suspected as such because `String#singularize` didn't have it.

> Not liking the #with_count method name for this. 

Not the best with naming but just wanted to start a conversation on the idea of it

> So we could just invent one, like #plugularize. But better if we can find an existing word that's closer. Then it can take the place of the #with_count method you have and we can drop the singularize(count) and deprecate pluralize(count).

Just an idea after a long day on Friday..
```ruby
# Integer#unit
1.unit('day') # 1 day
3.unit('day') # 3 days
1.unit('days') # 1 day
3.unit('days') # 3 days
```
Usually the reason why you want to pluralize or singularize is to use it in a sentence like "You have 3 unread messages" (`3.unit('unread message')`. This doesn't directly solve the use case but it indirectly solves most use cases (I imagine). Thoughts?

oojewale commented:
Since we know that countable nouns have singular or plural `forms`. I'm inclined to suggest `#formalize_by_count`

> I did a quick search and it doesn't seem like there's a single word that represents the process of either singularizing or pluralizing a word. So we could just invent one, like #plugularize. But better if we can find an existing word that's closer.

dhh commented:
I think part of the trouble here is thinking of this as something we have to cram onto Integer or String. I don't think that's necessary. We can simply keep this as a method on the Inflector and go with the clearest, most descriptive name: `Inflector#singularize_or_pluralize(word, count)`.

lugray commented:
The verb is _decline_ (https://en.wikipedia.org/wiki/Declension), but I don't think that would make things clearer.
